packages/kas/src/nodes.ts

```diff
@@ -69,26 +54,77 @@ var randomFloat = function (min, max) {
 var ITERATIONS = 12;
 var TOLERANCE = 9; // decimal places
 
+// NOTE(kevinb): _.partition exists in a more recent version of underscore.
+// To avoid having to update underscore I've added a hacky version of this
+// method here.
+function partition<T, V extends _.Collection<T>>(
+    list: V,
+    iteratee: _.CollectionIterator<T, boolean>,
+): [_.TypeOfCollection<V>[], _.TypeOfCollection<V>[]] {
+    const a: _.TypeOfCollection<V>[] = [];
+    const b: _.TypeOfCollection<V>[] = [];
+    _.forEach(list, (elem, key, ctx) => {
+        if (iteratee(elem, key, ctx)) {
+            a.push(elem);
+        } else {
+            b.push(elem);
+        }
+    });
+    return [a, b];
+}
+
+function isExpr(arg: string | Expr): arg is Expr {
+    return arg instanceof Expr;
+}
+
+function isRational(arg: Expr): arg is Rational {
+    return arg instanceof Rational;
+}
+
+function getFactors(expr: Expr): Expr[] {
+    if (expr instanceof Mul) {
+        return expr.terms;
+    } else {
+        return [expr];
+    }
+}
+
+type Hints = {
+    parens?: boolean;
+    divide?: boolean;
+    root?: boolean;
+    subtract?: boolean;
+    fraction?: boolean;
+    entered?: boolean;
+    negate?: boolean;
+    open?: boolean;
+};
+
+type Vars = Record<string, string>;
+
+type Options = {
+    preciseFloats?: boolean;
+    once?: boolean;
+};
+
 /* abstract base expression node */
-class Expr {
+abstract class Expr {
+    hints: Hints;
+
     constructor() {
         this.hints = {
             parens: false,
         };
     }
 
     // this node's immediate constructor
-    func() {
-        abstract();
-    }
+    abstract func: {new (...args: any[]): any; name: string};
```

Comment:
I'm having a hard time deciphering this line. Would it be worth a comment explanation or something?

Reply:
Will do.  The `new` part of the type let's TypeScript know that `func` should be a class.